pennylane/transforms/optimization.py

```diff
@@ -0,0 +1,244 @@
+            else:
+                if current_gate.is_symmetric_over_wires:
+                    list_copy.pop(next_gate_idx + 1)
+                else:
+                    current_gate.queue()
+        # Otherwise, queue and move on to the next item
+        else:
+            current_gate.queue()
+
+        # Remove this gate from the working list
+        list_copy.pop(0)
+
+    # Queue the measurements normally
+    for m in tape.measurements:
+        m.queue()
+
+
+@qfunc_transform
+def merge_rotations(tape):
+    """Quantum function transform to combine rotation gates of the same type
+    that act sequentially.
+
+    If the combination of two rotation produces an angle that is close to 0,
+    neither gate will be applied.
+
+    Args:
+        tape (.QuantumTape): A quantum tape.
+
+    **Example**
+
+    Consider the following quantum function.
+
+    .. code-block:: python
+
+        def qfunc(x, y, z):
+            qml.RX(x, wires=0)
+            qml.RX(y, wires=0)
+            qml.CNOT(wires=[1, 2])
+            qml.RY(y, wires=1)
+            qml.Hadamard(wires=2)
+            qml.CRZ(z, wires=[2, 0])
+            qml.RY(-y, wires=1)
+            return qml.expval(qml.PauliZ(0))
+
+    The circuit before optimization:
+
+    >>> dev = qml.device('default.qubit', wires=3)
+    >>> qnode = qml.QNode(qfunc, dev)
+    >>> print(qml.draw(qnode)(1, 2, 3))
+    0: ───RX(1)──RX(2)──────────╭RZ(3)──┤ ⟨Z⟩
+    1: ──╭C──────RY(2)──RY(-2)──│───────┤
+    2: ──╰X──────H──────────────╰C──────┤
+
+    By inspection, we can combine the two ``RX`` rotations on the first qubit.
+    On the second qubit, we have a cumulative angle of 0, and the gates will cancel.
+
+    >>> optimized_qfunc = merge_rotations(qfunc)
+    >>> optimized_qnode = qml.QNode(optimized_qfunc, dev)
+    >>> print(qml.draw(optimized_qnode)(1, 2, 3))
+    0: ───RX(3)─────╭RZ(3)──┤ ⟨Z⟩
+    1: ──╭C─────────│───────┤
+    2: ──╰X──────H──╰C──────┤
+    """
+    # Make a working copy of the list to traverse
+    list_copy = tape.operations.copy()
+
+    while len(list_copy) > 0:
+        current_gate = list_copy[0]
+
+        # Normally queue any non-rotation gates
+        if not current_gate.is_composable_rotation:
+            current_gate.queue()
+            list_copy.pop(0)
+            continue
+
+        # Find the next gate that acts on the same wires
+        next_gate_idx = _find_next_gate(current_gate.wires, list_copy[1:])
+
+        # If no such gate is found (either there simply is none, or there are other gates
+        # "in the way", queue the operation and move on
+        if next_gate_idx is None:
+            current_gate.queue()
+            list_copy.pop(0)
+            continue
+
+        # Otherwise, get the next gate
+        next_gate = list_copy[next_gate_idx + 1]
+
+        # If next gate is the same, we can potentially remove it
+        if current_gate.name == next_gate.name and current_gate.wires == next_gate.wires:
```

Comment:
This function should be further optimized using some sort of clever queuing or lookahead to catch cases where there are multiple rotations to accumulate together. Right now, something like 
```
RX(a)
RX(b)
RX(c)
```
would need 2 passes to turn to `RX(a+b+c)`, since it will first handle `RX(a)` and `RX(b)`, and then catch `RX(c)` on the next pass. It should instead be able to achieve this all in one go.